move `once` module out of `poison`

connortsui20 opened the pull request:
From https://github.com/rust-lang/rust/issues/134645#issuecomment-3324577500, since `Once` will not have a non-poisoning variant, we remove it from the `poison` module.

Additionally:

1. Renames `once::ExclusiveState` to `OnceExclusiveState` since it was a bit confusing reading just `ExclusiveState` where it is used.
2. Reorders a few module definitions and re-exports in `library/std/src/sync/mod.rs` for clarity.


Also, once this is merged, I think that we can begin the process of stabilizing [`sync_poison_mod`](https://github.com/rust-lang/rust/issues/134646) and [`sync_nonpoison`](https://github.com/rust-lang/rust/issues/134645)

tgross35 commented on library/std/src/sync/poison.rs:
```diff
@@ -59,18 +51,19 @@
+//!
+//! Note that the [`Once`] type also employs poisoning, but since it has non-poisoning `force`
+//! methods available on it, we do not need to both create a separate `nonpoison::Once` type and
+//! also place [`Once`] in this module.
```
Since this is user-facing the wording like "we need to create" is a bit odd, think this can be simplified

```suggestion
//! Note that the [`Once`] type also employs poisoning, but since it has non-poisoning `force`
//! methods available on it, there is no separate `nonpoison` and `poison` version.
```

tgross35 commented on library/std/src/sync/poison.rs:
```diff
@@ -13,8 +13,8 @@
-//! For the alternative implementations that do not employ poisoning,
-//! see [`std::sync::nonpoison`].
+//! The synchronization objects in this module have alternative implementations that do not employ
+//! poisoning on the [`std::sync::nonpoison`] module.
```
```suggestion
//! The synchronization objects in this module have alternative implementations that do not employ
//! poisoning in the [`std::sync::nonpoison`] module.
```

tgross35 commented on library/std/src/sync/mod.rs:
```diff
@@ -193,47 +203,42 @@ pub use self::once_lock::OnceLock;
     Condvar,
-    Once, OnceState,
     RwLock, RwLockReadGuard, RwLockWriteGuard,
 };
```
Looks like alphabetical ordering may have gotten flipped between `Mutex` and `TryLockError`. (would be nice if we could just remove the `rustfmt::skip` from this module but it does make things messy)

tgross35 reviewed (commented):
A few nits here then r=me

> Additionally:
> 
>     1. Renames `once::ExclusiveState` to `OnceExclusiveState` since it was a bit confusing reading just `ExclusiveState` where it is used.
>     2. Reorders a few module definitions and re-exports in `library/std/src/sync/mod.rs` for clarity.

For future reference, since these and the original `Once` change are all orthogonal, they'd ideally be split to three separate commits. It's a separation of concerns (cleanup vs. behavior change) and makes the diffs easier to follow.

*[View changes since this review](https://triagebot.infra.rust-lang.org/gh-changes-since/rust-lang/rust/147125/a00a5159cfac23e9b1551566cf71ff31edc1f11b..ef16d450d47ad871910a6098748db3aa6f1be7b4)*

connortsui20 committed:
rename `once::ExclusiveState` to `OnceExclusiveState`

It is a bit confusing when reading code that uses this type since it is
not immediately obvious that it is specific to `Once`.

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
clean up some documentation

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
reorganize `library/std/src/sync/mod.rs` file

Moves things around to make a bit more sense (plus prepare moving `once`
out of `poison`.

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>

connortsui20 committed:
move `once` module out of `poison`

Since `Once` will not have a non-poisoning variant, we remove it from
the `poison` module.

Signed-off-by: Connor Tsui <connor.tsui20@gmail.com>
